### ISSUE TITLE
make it available from a Maven repo

### ISSUE THREAD

**Issue by abstratt** [cjwizard/cjwizard#14]
Consider making this project available from a Maven repo.

**Comment by PhoneixS** [comment]
Only to keep it in mind, slf4j have a tool to help converting to its format. It can be found in http://www.slf4j.org/migrator.html.

**Comment by guerda** [comment]
Availability via Maven would be huge plus. I could support you in the transition phase and for testing purposes.

**Comment by guerda** [comment]
I forked this repo and without the log4j/commons-logging discussion, this project is ready for maven. I tested it against my own project.
Propably, you want to have a look at my branch/pull-request. It was very little work, so it could be merged easily without being dependent on #3.

**Comment by PhoneixS** [comment]
Ok, thank you for your effort. When I have more time I will test it and merge to the trunk.

**Comment by creswick** [comment]
Great!

Thanks Philip.

**Comment by creswick** [comment]
Gah, Github seems to be doing odd things with my comment edit.  To be clear:

@PhoneixS The pull request looks good to me -- I'd suggest merging it if you don't have objections.  +1 for travis integration, too.

**Comment by guerda** [comment]
Cool! I look into the maven publishing, too.

**Comment by PhoneixS** [comment]
Done, I have merged it.

I think that @creswick must add the project to travis as is the one with enough privileges in github.com/creswick/cjwizard repository (read more about it in http://docs.travis-ci.com/user/getting-started/) so the official repository is checked instead of Guerda's repository.

Also to be noted is that the project use maven but it isn't "mavenized" as the layout is the old layout. In maven normally you have the folders `src/main/java/com/...` instead of only `src/com/...` but it's a step on the right direction.

**Comment by guerda** [comment]
Thanks for the merge!
Regarding the standard layout for Maven projects: I wanted to create a running project which builds in Travis successfully before refactoring the source folder layout.
I will do this in a later pull request.
Regarding Travis CI: You are correct, the Travis webhook must be activated by the repo owner. Then the README.md must be updated. I cannot do this via a pull request.

**Comment by guerda** [comment]
Here's the PR: #30

**Comment by guerda** [comment]
Here are some basic information how to publish to OSSRH and then to Maven Central: http://central.sonatype.org/pages/ossrh-guide.html

Some of the requirements for the deployment:
* correct information (GAV)
* developer information including urls
* scm information
* GPG signed artifacts
* correct artifact naming
* JIRA ticket for Maven central

**Comment by creswick** [comment]
Merged the PR, and updated the travis url to point to my travis account's build (which I also just enabled).

I'm not sure when I can work through the OSSRH details; probably not this week, but I'll try to do so as soon as possible (possibly this weekend; the 7-8th).  If there are infrastructure details, such as generating signed releases, that you (@guerda or @PhoneixS) could help with, that'd be great!

**Comment by creswick** [comment]
Here's the OSSRH ticket:

https://issues.sonatype.org/browse/OSSRH-14241

**Comment by guerda** [comment]
@creswick I'm currently working on the correct format for the release plugin, so yet another PR will come soon. You just need a GPG key pair to perform the task afterwards.

**Comment by guerda** [comment]
#31 contains GPG signing, License information, sources jar inclusion and developer information. Should be all to release on OSSRH. Have fun!

**Comment by creswick** [comment]
Thanks!

So, unfortunately, my domain registration lapsed, and I don't have ciscavate.org anymore.  It looks like I might be able to get it back, but we may need to change the package/group prefix before the package will be accepted.

**Comment by PhoneixS** [comment]
It is a big change but I think that it is necessary. And maybe is good to make an official organization in github (can be also called cjwizard) so you can use com.github.cjwizard as package an translate the "oficial" repository to that organization.

**Comment by creswick** [comment]
The github organization sounds good to me.  I've created one and forked the repo over there.

Does anyone know if it's possible to move the tickets? https://github.com/cjwizard/cjwizard

**Comment by PhoneixS** [comment]
@creswick you don't need to fork it, you must transfer the owner of the repository to the group. This way issues and other things are also transferred.

I have deleted the fork so you can transfer it without conflict.

**Comment by creswick** [comment]
> I have deleted the fork so you can transfer it without conflict.

Thanks!

**Comment by creswick** [comment]
I created a branch for the refactor, switching the groupId to 'com.github.cjwizard', updating the metadata in the pom.xml, and changing the 'org.ciscavate' prefix to 'com.github.cjwizard' in the packages.

The branch is here: https://github.com/cjwizard/cjwizard/tree/githubGroupId

Could you folks (@PhoneixS and @guerda) try that branch and let me know if I broke anything? The tests pass, but I'm not confident that they cover much...

I've bumped the version to 0.22, also, but that may need to go higher (1.0?) to indicate the breaking change.

**Comment by PhoneixS** [comment]
Yes I also think that is better to increase to 1 because it had changed the layout and many other things.

**Comment by PhoneixS** [comment]
Also you need to change the Travis CI repository so it point to the correct repository (although it is running now by redirect of github).

**Comment by creswick** [comment]
Bumped version to 1.0.0-snapshot.

I think I'm waiting on Travis to see the cjwizard org (I did just authorize access) before I can update / create a new travis job.

**Comment by guerda** [comment]
@creswick  For me, it looks fine. We should definitely work on the unit tests, but that's a second step, imho. I think you can merge it to master and publish a new release 1.0.

**Comment by PhoneixS** [comment]
As it is already merged I've just updated the Travis CI repository to the new organization repository (https://travis-ci.org/cjwizard/cjwizard).

**Comment by boformer** [comment]
Any news on this?

**Comment by creswick** [comment]
I had some issues with the maven package commands, and may need help getting that straightened out, but I haven't had time to sit down and write up a ticket.

**Comment by Petikoch** [comment]
Hmmm... this is still open :-(

It seems that you had some issues to bring it to "maven central"? As workaround... publishing it to Bintray (a competitor repo) is IMO simpler... Check out e.g. https://blog.bintray.com/2015/09/17/publishing-your-maven-project-to-bintray/ 

Best regards,
Peti

**Comment by creswick** [comment]
@Petikoch I had some issues with the maven process; rooted in my general lack of time to dedicate to this project :(

I'm happy to help transition if anyone else is interested and willing to take over as maintainer!

**Comment by creswick** [comment]
@Petikoch Bintray doesn't seem to be working for me, unfortunately.  Any suggestions on this failure?

```bash
[ERROR] Failed to execute goal org.sonatype.plugins:nexus-staging-maven-plugin:1.6.3:deploy (injected-nexus-deploy) on project cjwizard: Failed to deploy artifacts: Could not transfer artifact com.github.cjwizard:cjwizard:jar.asc:sources:1.0.3-20160810.225152-1 from/to bintray-repo-cjwizard (https://api.bintray.com/maven/creswick/CJWizard/cjwizard/;publish=1): Failed to transfer file: https://api.bintray.com/maven/creswick/CJWizard/cjwizard/;publish=1/com/github/cjwizard/cjwizard/1.0.3-SNAPSHOT/cjwizard-1.0.3-20160810.225152-1-sources.jar.asc. Return code is: 400, ReasonPhrase: Bad Request. -> [Help 1]
```

**Comment by Petikoch** [comment]
@creswick , I don't know what's the source of the problem above. 

Maybe there is an issue with the repo ID? According to the Bintray UserManual https://bintray.com/docs/usermanual/formats/formats_mavenrepositories.html#anchorMavenDeploy it has the form _bintray-[YOUR_USERNAME]-[REPOSITORY_NAME]_

Therefore the section in the pom.xml should probably look like

    <distributionManagement>
      <repository>
        <id>bintray-creswick-CJWizard</id>
        <url>https://api.bintray.com/maven/creswick/CJWizard/cjwizard/;publish=1</url>
      </repository>
    </distributionManagement>

I hope this helps.

Best regards,
Peti

**Comment by Petikoch** [comment]
@creswick 

After the sucessful publishing to Bintray, you can manually add your package to "JCenter" (https://bintray.com/bintray/jcenter). Bintray "JCenter" is something like maven central.

**Comment by creswick** [comment]
@Petikoch Thanks - the repository ID is documented as an arbitrary string; it just needs to match the corresponding authentication section in my Maven settings.xml (which is obviously not published).

I did change that, and it may have helped.  The other thing that I didn't realize is that you can't upload SNAPSHOT versions, so changing to 1.0.4 seems to have worked (although the deploy command *did* fail...)

Could you try this out and let me know if it worked?

Here's the bintray release url:

https://bintray.com/creswick/CJWizard/cjwizard/1.0.4

**Comment by Petikoch** [comment]
@creswick , can you add it to jcenter? https://bintray.com/bintray/jcenter ("Include my package")

Usually people use either maven-central or jcenter or both.

Congrats for the release!

Best regards,
Peti

**Comment by creswick** [comment]
That *should* be underway now... I've had a number of failures with mvn
deploy, but I think that finally worked (on v. 1.0.5) Hopefully it'll be
transitioned to JCenter soon. (I just filed another request.)

**Comment by Petikoch** [comment]
Hi @creswick,

congrats, it's in Bintray's jcenter!

Best regards,
Peti

**Comment by Petikoch** [comment]
I think this issue could be closed now...

**Comment by creswick** [comment]
Closing; 

For reference: CJWizard is up on JCenter here: https://bintray.com/creswick/CJWizard/cjwizard

**Comment by Petikoch** [comment]
For reference: It's now here: https://bintray.com/cjwizard/CJWizard/cjwizard
